### PR TITLE
Add spell check TextSpan creation logic that doesn't rely on composing region

### LINKED ISSUE #119534: [Android][iOS] Ensure misspelled text is only highlighted once the word is not being composed
Currently, spell check logic [uses the composing region](https://github.com/flutter/flutter/blob/df8ad3d2cb25e658c3600ae6fa0931404c1a8ed1/packages/flutter/lib/src/widgets/spell_check.dart#L249) to determine whether a misspelled word is being composed and thus, whether or not it should be highlighted as misspelled. This approach is flawed because iOS and Android devices not using Gboard may not have a valid composing region to rely on. Therefore, this logic must be refactored as to not be reliant on the composing region.

### PR DESCRIPTION
**Pull request by camsim99**
Fixes https://github.com/flutter/flutter/issues/119534 by adding extra logic to allow the proper `TextSpan` trees to be build with spell check results that does not rely on the composing region.

Also, refreshes the algorithm used to correct spell check results to improve cases where spell check results are out of date, but the composing region line does not help smooth over this fact by giving the framework extra time to update while a word is being composed (because a composing region line would cover up any flashing red underlines).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:303]: Split each parameter onto its own line.

**Review by justinmc — COMMENTED** [review]
Overall the approach in _buildSubtreesWithoutComposingRegion looks good to me, but two things I'm wondering about:

  1. Is it also responsible for drawing the composing region?  Say if you use a Chinese composing IME on Android with Gboard, will it get routed through _buildSubtreesWithoutComposingRegion and the composing region won't draw?  Or maybe this is handled somewhere and there's nothing to worry about.
  2. Depending on the answer to the above, it feels like we might be able to share a lot of the code between _buildSubtreesWithComposingRegion and _buildSubtreesWithoutComposingRegion.  It seems like the fundamental difference between these two methods is just how we detect the word that's currently being edited.  Maybe there is more about drawing the composing region that needs to be different, though.

**Comment by camsim99** [comment]
> 1. Is it also responsible for drawing the composing region?  Say if you use a Chinese composing IME on Android with Gboard, will it get routed through _buildSubtreesWithoutComposingRegion and the composing region won't draw?  Or maybe this is handled somewhere and there's nothing to worry about.
> 2. Depending on the answer to the above, it feels like we might be able to share a lot of the code between _buildSubtreesWithComposingRegion and _buildSubtreesWithoutComposingRegion.  It seems like the fundamental difference between these two methods is just how we detect the word that's currently being edited.  Maybe there is more about drawing the composing region that needs to be different, though.

It's not responsible for drawing the composing region at all, but it does ignore the word the cursor is touching. However, if it's running on an Android device, if won't call that method, so the problem you described in # 1 should not be a problem, unless I'm misunderstanding what you're asking.

Nonetheless, we can probably share some code between the two, although detecting the cursor index to ignore the misspelled word is much simpler logic than dealing with the composing region, so I think some readability would be lost. I would rather not spend time on this now though since I have limited time and I want to focus on getting this landed in time, unless you think it's high priority. Just lmk!

**Comment by justinmc** [comment]
Sounds good, as long as the composing region is being drawn the the code reuse thing is secondary.

**Comment by camsim99** [comment]
> Sounds good, as long as the composing region is being drawn the the code reuse thing is secondary.

Okay. Does TargetPlatform.android pick up Samsung?

**Comment by justinmc** [comment]
Yes, a Samsung Android device will have `defaultTargetPlatform == TargetPlatform.android`.  However, there is no way in the framework to know if it is specifically Samsung or not.

**Comment by camsim99** [comment]
> Yes, a Samsung Android device will have `defaultTargetPlatform == TargetPlatform.android`. However, there is no way in the framework to know if it is specifically Samsung or not.

Okay, gotcha. I will make sure to test this on one today.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:241]: I just tried it: It works the same, but the composing region doesn't draw. I think that's expected behavior!

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:225]: I asked about `TargetPlatform.android` including Samsung because I don't understand why this would crash? Does the Samsung keyboard not have a composing region? If so, I didn't realize that.

I still need to test this, but it should not be a problem because in addition to this check, I can just check if the composing region is valid before using the method that considers it.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [inline packages/flutter/lib/src/widgets/spell_check.dart:187]: The one thing I still want to think about a little more is whether or not to calculate offset this way. The rest of the method assumes really only that one modification has been made between `resultsText` and `newText`, and while it may account for more cases than that, this explicitly allows for two operations: modification and movement.

Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:187]: Thinking about this more, this algorithm covers the following cases:

 1. One or more `SuggestionSpan` shifts, e.g. "Hlello, wrold" --> "Hlello, beautiful wrold" or --> "Oh! Hlello, big, beautiful wrold"
 2. One or more `SuggestionSpan` modifications, e.g. "Hello, wrold" --> "Hello, wroldd" or --> "Helllo, wroldd"
 3. One or more `SuggestionSpan` shifts or modifications, e.g. "Hello, wrold" --> "Hello, big wroldd"

This should be all of the cases that we need to consider because we only care about finding the `SuggestionSpan` text coming from the old text (`resultsText`) in the new text, which means that we only care about shifts and modifications of that `SuggestionSpan` text. Calculating the offset this way in this case, means that we catch case 2 and 3. Case 1 is caught above in the code by adjusting `offset` based on the actual `foundIndex` of the `SuggestionSpan` text. Thus, the code does not at all assume that only one modification has been made, and I believe it is fine as is. I still think multiple modifications/shifts is unlikely, but it doesn't hurt to cover that case.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:103]: I think the indentation here is wrong.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:233]: It does have a composing region, it just doesn't use it for the current editing word (as far as I know).  Maybe I'm too pessimistic.  I thought the Samsung keyboard wasn't compatible at the engine level or something.  If it works then that's great.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:249]: That makes me think that we could use only this method and get rid of the other one.  It could remove the composing region when it's part of the replaced word and then draw it.  Maybe in another PR if that's going to delay this one, though.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:107]: Why did you remove this comment?  I think we need something explaining what _correctSpellCheckResults is used for.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:187]: How are you able to do cases 2 and 3?  If I'm understanding correctly, those show that a misspelled word has been modified (`wrold` to `wroldd`), and you're still able to correct it (`world`)?  That seems super tricky if I'm understanding it correctly.  It might be better to not handle those at all.  What if the modification creates a correctly spelled word (`worl` to `worlds`)?

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:198]: Any thoughts on the performance of this algorithm?  It seems like it repeatedly uses `substring` in order to reduce the search space (I think substring would be O(1) and therefore a valid optimization) and then `indexOf` to search, which I would think is O(reduced search space) in the worst case.  Probably reasonable given that the misspelled words could have moved anywhere in the string, but let me know if I'm missing anything.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:107]: This was an accident! Thanks!

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:249]: I definitely think that they can be merged into one and is worth pursuing, then. I am running short on time, though, so unless you're open to playing around with it, I say we delay it for now and file an issue!

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:198]: I do believe that's right and I had the same concern. I think some optimizations could be made if we make further assumptions about when this method will be needed, e.g. if we assume only one modification was made, then we'd only need to do one search and then use lookups to verify the rest of the results based on the first (and only, in this case) calculated `offset` and `searchStart`. I think it is worth considering, my only concern here is time.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:187]: Not quite -- case 2 means that a word like `wrold` has been modified to `wroldd`, and it would correct the 
`SuggestionSpan` of that word from `SuggestionSpan(TextRange(n, n+5), <String>["world"])` to  `SuggestionSpan(TextRange(n, n+6), <String>["world"])`, for example. It's this case (and 3 because 3 involves 2) that pushed me to change this method because if you are typing faster than the spell check results update, the red underline flashes under the misspelled word because of the lag (the cursor index/base offset would not be considered as touching the misspelled word you are composing because an "unknown" addition to the word has been made). My changes fix this issue in the case you described. You would still be compoinsg `worl` --> `worlds`, but none of it would be underlined in red, unless after composing, it is still misspelled.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:204]: Sounds good, it should be fine as-is for now.  We haven't done a lot of work on optimizing for large amounts of text in general in Flutter yet.

Also we probably couldn't ever assume that there was only one modification, even though that's probably the case 99% of the time (because the user pressed a single key on their keyboard).  It's possible that the IME has sent us any wild modification or even an entirely new string.  We encountered the same problem in @Renzo-Olivares's deltas work.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:193]: Ok I think I got it, we definitely do want the ability to tell that the word is still being edited in these cases so we can hide the red underline.

**Comment by justinmc** [comment]
Alright now I've tested it on the Android emulator too.  None of the problems above happen on Android (Gboard).  

  1. Type two misspelled words separated by a semicolon.  Replace the second word.  Sometimes, extra characters will remain after the replacement at the end of the new word.


https://user-images.githubusercontent.com/389558/229881701-8e18649c-194d-40da-bedd-7507029a750c.mov

  2. Sometimes, the cursor does not return to the end of the word after tapping a replacement.


https://user-images.githubusercontent.com/389558/229882182-cffd30b9-b102-47b5-b41c-898b088157c5.mov

**Comment by camsim99** [comment]
@justinmc For iOS:

1/2: These are valid issues because the code relies on spaces a lot. I'll take a look today, but worst case, let's file bugs before  this makes it to stable.

3: This is expected! Although, iOS must have done an update, because I now see that for misspelled words, the blue highlight is red. May need to file an issue for this before this reaches stable, as well.

For Android:

1: This is probably an issue on the framework side, too. I echo what I said about filing an issue!

2: I'm not 100% sure I knew it was supposed to do this. I do see it on native, though. It's a simple fix, so I'll also look at that today.

**Comment by justinmc** [comment]
@camsim99 Are you aware that on master spell check suggestions sometimes appear blank on iOS?

Screenshot 2023-04-04 at 11 57 44 AM

No worries because it's fixed in this PR.  However, that's a strong reason to make sure this PR gets shipped.

**Comment by justinmc** [comment]
I tried to reproduce all of the numbered bugs I posted above again on master, and they all either can still be reproduced or are blocked by the "blank suggestion" bug I screenshotted.  So, this PR does not appear to introduce any new bugs and is an improvement across the board.  I agree that we should just file issue for all of the bugs above and move forward with this PR.

No worries about fixing Android bug number 2 here if it's not super simple.

**Comment by camsim99** [comment]
> Are you aware that on master spell check suggestions sometimes appear blank on iOS?

Yes, this is because you are using `TextField` versus `CupertinoTextField` :/  With Cupertino style menu, it will say "no replacement suggestions found" but it doesn't do that in the Material style menu. Thus, this PR won't fix it but we know this is an issue and I will file an issue for that, too.

**Comment by justinmc** [comment]
Oh right, I keep forgetting that.  Thanks!

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:462]: Good find with all the commas!

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/spell_check_test.dart:22]: Why make all of these `testWidgets` instead of `test`?  I think if it doesn't use `tester` it should just be a `test`.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/spell_check_test.dart:47]: Did these tests fail or you just changed them to mobile-only since that's what spell check supports right now?

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/spell_check_test.dart:153]: Is this case with a composing region tested somewhere else?  If not, maybe we should keep both, one original with the composing region and one your update without composing region?

Same for other tests where the composing region was removed.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/spell_check_test.dart:353]: This parenthesis should be indented.

**Review by justinmc — COMMENTED** [review]
Minor nits on the tests, otherwise looks good.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (camsim99) [reply packages/flutter/lib/src/widgets/spell_check.dart:420]: Learning slowly!! :)

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (camsim99) [reply packages/flutter/test/widgets/spell_check_test.dart:22]: I did this to use `TargetPlatformVariant`

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (camsim99) [reply packages/flutter/test/widgets/spell_check_test.dart:47]: I mean we can have them run on all, but mobile is what matters so I'd figure I'd change it just to be sure we are testing what we support. Your call! I need them to at least run on those too to try the two different ways of building the `TextSpan` tree with spell check results.

**Comment by camsim99** [comment]
> @justinmc For iOS:
> 
> 1/2: These are valid issues because the code relies on spaces a lot. I'll take a look today, but worst case, let's file bugs before this makes it to stable.
> 
> 3: This is expected! Although, iOS must have done an update, because I now see that for misspelled words, the blue highlight is red. May need to file an issue for this before this reaches stable, as well.
> 
> For Android:
> 
> 1: This is probably an issue on the framework side, too. I echo what I said about filing an issue!
> 
> 2: I'm not 100% sure I knew it was supposed to do this. I do see it on native, though. It's a simple fix, so I'll also look at that today.

Just reverted logic to extend `SuggestionSpan`s when misspelled words have been modified, thanks to @justinmc's suggestion. To update on these issues, based on what I now observe:

For iOS,

1/2: Seems like these are a direct result of the actual spell check results we get back from iOS. Respectively, they do seem to include the semicolon and do not seem to include the apostrophe in the results (at least in that moment while composing the word, e.g "hlelo'" does not have results that include the apostrophe but "hlelo's" does). May have to correct this on the framework end.

3: Native behavior.

For Android,

1: This revert fixes this issue!

2: This is still an issue we will need to address on the framework side.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (camsim99) [reply packages/flutter/test/widgets/spell_check_test.dart:153]: I am open to this, but I don't think it's necessary. Cases with the composing region are tested in the other tests, and the purpose of the tests where I removed it have to do with correcting the spell check results versus correctly styling the composing region.

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:143]: Nit: I would format it like this:

```dart
      final SuggestionSpan adjustedSpan = SuggestionSpan(
        TextRange(
          start: currentSpan.range.start + offset,
          end: currentSpan.range.end + offset,
        ),
        currentSpan.suggestions,
      );
```

File: packages/flutter/lib/src/widgets/spell_check.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/spell_check.dart:154]: Nit: Outdent these two lines.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [reply packages/flutter/test/widgets/spell_check_test.dart:22]: Ah I didn't realize you couldn't use it with `test`!  Sounds good then.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [reply packages/flutter/test/widgets/spell_check_test.dart:47]: I say keep it as you have it 👍 .  I've seen unnecessary failures happen because a test was running on a platform that it was irrelevant for.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [reply packages/flutter/test/widgets/spell_check_test.dart:153]: Ok sounds good to keep it like this if it's covered elsewhere 👍

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/spell_check_test.dart:328]: What happened to the test that you added in 7904ab2cc4d112071c2a65d16a54d845399a058f?

**Review by justinmc — APPROVED** [review]
LGTM 👍.  Just a few nits.

File: packages/flutter/test/widgets/spell_check_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/spell_check_test.dart:330]: Nit: Maybe update this name?  Since it actually throws away the spell check suggestion now.

**Comment by camsim99** [comment]
Frob is having issues and the branch cut is today. Merging to get this in prior.

**Pull request merged**
